drivers/i2c/i2c_mchp_xec.c

```diff
@@ -277,9 +396,98 @@ static int i2c_xec_configure(const struct device *dev,
 		return -EINVAL;
 	}
 
-	i2c_xec_reset_config(dev);
+	return i2c_xec_reset_config(dev);
+}
 
-	return 0;
+/*
+ * Attempt to recover the I2C bus.
+ * If SCL is driven low by an external device all we can do
+ * is poll a few time and hope the external device will release SCL.
+ * If SDA is driven low by an external device we will generate 9
+ * clocks and attempt to generate a STOP. Sometimes this can trigger
+ * the stuck device to release SDA.
+ *
+ * We can make use of the controller's bit-bang mode because we aren't
+ * performing a real I2C transaction.
+ * Switch to big-bang undriven mode: our drives tri-stated.
+ * Get SDA and SCL pin states.
+ * If SCL is low
+ *   Retry SCL every 5 us up to I2C_RECOVER_SCL_LOW_RETRIES times.
+ *   If SCL remains low reset controller and return -EBUSY
+ * If SDA is low
+ *  Retry driving 9 clocks at ~100 KHz and generate a STOP up
+ *  to I2C_RECOVER_SDA_LOW_RETRIES
+ *  If SDA remains low reset controller and return -EBUSY
+ * reset controller
+ * return success
+ */
+static int i2c_xec_recover_bus(const struct device *dev)
+{
+	const struct i2c_xec_config *config =
+		(const struct i2c_xec_config *const) (dev->config);
+	struct i2c_xec_data *data =
+		(struct i2c_xec_data *const) (dev->data);
+
+	uint32_t ba = config->base_addr;
+	uint32_t cfg;
+	int count;
+
+	data->started = 0;
+	data->pending_stop = 0;
+	data->slave_read = 0;
+
+
+	/* Make sure our controller is idle before attempting bus recovery */
+	cfg = MCHP_I2C_SMB_CFG(ba);
+	MCHP_I2C_SMB_CFG(ba) = MCHP_I2C_SMB_CFG_RESET;
+	xec_i2c_baud_clk_delay(ba, 16);
+	MCHP_I2C_SMB_CFG(ba) =
+		(config->port_sel & MCHP_I2C_SMB_CFG_PORT_SEL_MASK);
+	MCHP_I2C_SMB_CFG(ba) |= MCHP_I2C_SMB_CFG_FEN;
+	MCHP_I2C_SMB_CFG(ba) |= MCHP_I2C_SMB_CFG_ENAB;
+
+	/* Enable bit-bang mode. */
+	MCHP_I2C_SMB_BB_CTRL(ba) = BIT(0);
+
+	count = I2C_RECOVER_SCL_LOW_RETRIES;
+	while (!(MCHP_I2C_SMB_BB_CTRL(ba) & BIT(5))) {
+		/* SCL is low. All we can do is wait for it to go high */
+		if (count-- == 0) {
+			/* reset controller which disables bit-bang mode */
+			i2c_xec_reset_config(dev);
```

Comment:
should we disable bitbag on error before reset_config() ?

Reply:
Controller reset disables big-bang